Send_position_target_global_int: Set mavlink altitude frame to be target's frame

hendjoshsr71 opened the pull request:
This changes the `send_position_target_global_int` message to send the target's altitude mavlink frame in the same frame as the target's location frame.

Note this currently ends up being filled by AC_WP_Nav::get_wp_nav() which always sets the target to ASML. 

**Master**
![master_set_altitude_frame_position_target_global_int](https://user-images.githubusercontent.com/69225461/116957556-1d1ccb00-ac66-11eb-9045-80a6ac1b7a8d.png)

**After PR**
![set_altitude_frame_position_target_global_int](https://user-images.githubusercontent.com/69225461/116957566-260d9c80-ac66-11eb-984d-da420d90234b.png)

khancyr commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -97,15 +97,30 @@ void GCS_MAVLINK_Copter::send_position_target_global_int()
+    case Location::AltFrame::ABOVE_HOME:
+        frame = MAV_FRAME_GLOBAL_RELATIVE_ALT;
+        break;
+    case Location::AltFrame::ABOVE_ORIGIN:
```
Hummm this one don't seems right. I will look what it should be

hendjoshsr71 replied:
Yeah. This isn't 100% as there is no equivalent mavlink_frame for the ekf origin. Really what should eventually be done is to ensure that the calling modes set the altitude frame according to the current commanded waypoint's frame.

khancyr replied:
MAV_FRAME_LOCAL_NED

khancyr replied:
but according to mavlink specs, valid frame are : 
Valid options are: MAV_FRAME_GLOBAL_INT = 5, MAV_FRAME_GLOBAL_RELATIVE_ALT_INT = 6, MAV_FRAME_GLOBAL_TERRAIN_ALT_INT = 11

khancyr replied:
So I would say, if that is in the specs, use the right frame otherwise default to MAV_FRAME_GLOBAL_INT even if that is a change in behavior

hendjoshsr71 replied:
Yeah. That is why I defaulted EKF_origin to become MAV_FRAME_GLOBAL as is currently being done. From my reading it seems that *_INT part has been dropped since it is redundant.

peterbarker replied:
This should be `location_alt_frame_to_mavlink_coordinate_frame`, like we have `bool GCS_MAVLINK::mavlink_coordinate_frame_to_location_alt_frame(const MAV_FRAME coordinate_frame, Location::AltFrame &frame)`

Even if it is a bit of a mouthful.

That won't work if we really do need to change the number, of course.

hendjoshsr71 replied:
Added `location_alt_frame_to_mavlink_coordinate_frame` to GCS_MAVLINK.

Is the comment of, "That won't work if we really do need to change the number, of course.", about the *_INT of the mavlink frame?

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -5300,6 +5300,31 @@ bool GCS_MAVLINK::mavlink_coordinate_frame_to_location_alt_frame(const MAV_FRAME
+        coordinate_frame = MAV_FRAME_GLOBAL_RELATIVE_ALT;
+        break;
+    case Location::AltFrame::ABOVE_ORIGIN:
+    default:
```
Remove this

hendjoshsr71 replied:
Removed

rmackay9 commented:
I think this is probably most useful in Guided mode where the caller may want to check that what they've sent in has been consumed correctly.

hendjoshsr71 commented:
- Based on a request in discord, added send_position_target_global_int to Sub
- moved the message to gcs_common
- TODO: add the autotests

**ArduSub receiving the message**
![image](https://user-images.githubusercontent.com/69225461/124874966-593a1800-df96-11eb-8677-3d8bca67912a.png)

hendjoshsr71 commented:
Updated this so that the commanded altitude frames for guided and auto modes can also report relative to home in addition to terrain & absolute frames.

Rover Autotest may fail: I think this is due to the issue with terrain in CI as it passes locally. There is an argument to be made that there isn't a lot of purpose to testing rover for differing altitude frames since it ignores altitude.  

Mission Below: Sets Relative Frame 100m then 150m, Terrain 100m then 150m, then Absolute Frame 1500m & 1550m `Position_Target_Global_Int` in red 
Mission switches
![set_altitude_frame_position_target_global_int](https://user-images.githubusercontent.com/69225461/131979034-67fe57e8-70b0-412d-86c6-20269415c70f.png)

hendjoshsr71 commented:
I've disabled the terrain frame autotest for now in this PR after fixing the altitude frame conversion in autotest. 

Mission Below: Sets Relative Frame 50m then 100m then 50m, Terrain 100m then 150m then 100m, then Absolute Frame 1200m then 1250m

Master: Red `Position_Target.Altitude`
![master_pos_target_alt_Frames](https://user-images.githubusercontent.com/69225461/135055536-bc385bd2-aaf4-48f7-a090-de2504adf0a5.png)

After PR: Green `Position_Target.Altitude`
![after_pos_target_alt_Frames](https://user-images.githubusercontent.com/69225461/135055569-fe21dc8b-fc45-4245-b033-ab133104dc1c.png)

hendjoshsr71 commented:
@peterbarker  Any thoughts on disabling `set_position_target_zzz` terrain frame testing as I do in this PR because of issue #18788 ? Or on my autotest changes here?
